### DISCUSSION TITLE
"Derived" Types from Primitive Value Types

### DISCUSSION THREAD

**Discussion by fitdev** [dotnet/csharplang#1583] [Language Ideas]
Perhaps the following is better addressed as a Roslyn issue than a language feature, but I am posting it here anyways.

It would be really nice to be able to create new domain-specific types based on / derived from primitive numeric types (in particular a _subset_ of their default range). For example, one can imagine a type `Probability` which would be "derived from" `double`, but have a range of `[0.0 - 1.0]`.

Currently one can implement desired functionality using the following techniques:
- Use `readonly struct`s with a single private backing field of the desired type (`double` in my example)
- Apply `StructLayout(LayoutKind.Explicit)` attribute
- Implement `public static implicit operator`s for automatic type conversion to/from base type

While this approach certainly works, it has several major drawbacks:
- It requires considerable amount of boilerplate code
- It has a significant performance penalty of being around x3-x5 times slower than if the base type was used directly (tried different variations, but it looks like conversion operators are not being JIT-inlined, even if all they do is set or get the backing field)

So, hopefully this will be addressed in some form or another, because there are many instances when the intent of the code as well as maintainability and readability can be made a lot clearer and easier if instead of general types like `double`, more specific user-defined types like `Percentage` or `Probability` could be used instead without incurring performance penalties, since underneath they are represented by the same basic system type.

**Comment by Joe4evr** [comment]
Duplicate of #524. Please search for your proposals first.

**Comment by fitdev** [comment]
@Joe4evr Thank you for pointing this out. I don't think it's a duplicate though, albeit very related to the issue you mentioned. My proposal is strictly about essentially being able to give an alias to an existing primitive type, that will be treated as different type, and for which you could define operators and static fields, but not much more than that, not a full blown inheritance by any means. Also emphasis here is on the performance, so that by the time code is JITed it will effectively be indistinguishable from using the actual primitive type like int.

**Comment by ufcpp** [comment]
So you mean https://github.com/dotnet/csharplang/issues/410, https://github.com/dotnet/csharplang/issues/341 ?

**Comment by fitdev** [comment]
@ufcpp More like an intersection of #524 and #410  but with emphasis on primitive types and performance. So it's not about extending those, rather just effectively creating new aliases for those that will be treated by the compiler / intellisense as distinct types, but as far as runtime is concerned those will remain their underlying types. Very similar to how enums are implemented I guess. And also, much like you can do with enums, you should be able to declare `const`ants for such aliased types (which is not true of structs in general).

**Comment by ufcpp** [comment]
> with emphasis on primitive types and performance

see also https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/first-class-structs.md

**Comment by bondsbw** [comment]
More discussion at #413.
